### PR TITLE
Change coefficient in piecewise interpolation functions and materials to pp

### PR DESCRIPTION
**Pull request by snschune**
see #15321  for description.

### CODE REVIEW

**Comment by snschune** [comment]
This PR is mainly for enabling easier use of `Piecewise<>` with stochastic tools 
so tag @aeslaughter and @zachmprince for their opinion. Here are the options to 
get the tests pass:
1. Submit patch to bison to fix failing actions. @gambka what do you think of these changes?
2. Add an additional parameter `postprocessor_scaling`.
3. Instead of using post processors make the scale factor a controllable Real and control it that way. @aeslaughter does stochastic tools work with controllable Reals well and/or is there a plan to do that.

**Comment by zachmprince** [comment]
@snschune, stochastic tools has two ways of transferring parameter values. First is using `MultiAppCommandLineControl`, which doesn't require the parameters to be controllable and simply uses command line syntax. The second is using `SamplerReceiver`, which does require that the parameter be controllable, but is capable of using the batch-restore system. Using the batch-restore system is much more efficient in both memory and runtime.

Right now, only controllable Real parameters can be used with `SamplerReceiver`.  So with your changes, it will not be able transfer parameter values to `scale_factor`. 

@aeslaughter, can you make sure what I said is correct?

**Comment by gambka** [comment]
@snschune I would be fine with these changes.  The problem is, if you want a single constant you would have to pass a `ConstantFunction` to a `FunctionValuePostprocessor` and then use that postprocessor value in the `scale_factor`. In actions this can be automated. I do like the flexibility of being able to use a postprocessor value but I recommend if you can to deprecate the old way for now to give time for downstream apps to update to the new approach. It might need to be in a two step approach where you use `addDeprecatedParam<Real>("scale_factor")` and then have something `addParam<PostprocessorName>("scale_factor_pp")`.  Then once the deprecated param is removed you just rename the input syntax of scale_factor_pp to scale_factor. I think the use of deprecation is more important for our external users that may not be updating their inputs that frequently. I wouldn't want something to just break for them.  

 I have also seen other objects such as `BodyForce` which supports a Real, Postprocessor, and FunctionName that are all multiplied together to create the scale factor. See: 

```
  params.addParam<Real>("value", 1.0, "Coefficient to multiply by the body force term");
  params.addParam<FunctionName>("function", "1", "A function that describes the body force");
  params.addParam<PostprocessorName>(
      "postprocessor", 1, "A postprocessor whose value is multiplied by the body force");
```
Then:

`factor = _scale * _postprocessor * _function.value(_t, _q_point[_qp]);`

 Is this an option to consider? Support both options?

**Comment by snschune** [comment]
From what @zachmprince said, I think the least invasive way is also the one that works the best.

The `BodyForce` approach that @gambka mentioned would be nice to have throughout for "RHS" kernels, AuxKernels, Functions etc. I don't think I want to add it to `Piecewise<>` right now because there is a less intrusive way.

**Comment by aeslaughter** [comment]
@zachmprince SamplerReceiver also works with std::vector<Real> parameters

**Comment by aeslaughter** [comment]
I am not sure what is happening in the PR. The change shown should have zero impact on how the code performs. The title mentions changing the scale factor to a PP, is that still in the works? 

As is, if you want to control the "scale_factor" for multiple objects it is possible to do that using "control_tags". There are some tests that show how to use it, these tags should work via the SamplerReciever as well but I am not sure anyone has tried it yet.

If you change the "scale_factor" to a PP it is possible still to connect to the Control system. You will have to create a ConstantPP that has a controllable "value" parameter. Then you just control PP rather than these objects.

Also, see my comment here: https://github.com/snschune/moose/commit/af25c467e10cd2887389d9a7c2b9d39ecf1d8ce7#r39323070

**Comment by aeslaughter** [comment]
@zachmprince This can be 4 of 5 for your CCB qualification.

**Comment by snschune** [comment]
@aeslaughter Zach helped me better understand the system. I thought I control via PP values but it's much easier to do that via controllable values. This new PR is much more light-weight and passes tests.

**Review by zachmprince — COMMENTED** [review]
@snschune could you add a test for this? I realized that we aren't testing our controllable parameters very well. You can probably base it off of test/tests/controls/real_function_control.

File: test/tests/controls/control_piecewise/controlled_piecewise.i [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+[]
+
+[Controls]
+  [./func_control]
+    type = RealFunctionControl
+    parameter = '*/*/scale_factor'
+    function = '2'
+    execute_on = 'initial'
+  [../]
+[]
```
Comment (zachmprince) [inline test/tests/controls/control_piecewise/controlled_piecewise.i:51]: Thank you!

**Review by zachmprince — DISMISSED** [review]

**Comment by snschune** [comment]
These failures are unrelated. Can this go in? @aeslaughter

**Comment by zachmprince** [comment]
@snschune what happened to the test?

**Comment by snschune** [comment]
I might have pushed form the wrong computer. It's been so long that I forgot where the commit is. 
Really, this is a two line change. How does it take months to get this merged?

**Review by zachmprince — DISMISSED** [review]

**Comment by snschune** [comment]
There it is, was on my laptop and I didn't check.

**Review by zachmprince — DISMISSED** [review]

**Review by loganharbour — DISMISSED** [review]
Months?! 3.5 weeks :-)

This'll get in once it goes green again

File: framework/src/materials/PiecewiseLinearInterpolationMaterial.C [hunk 2]
```diff
@@ -37,6 +37,7 @@ PiecewiseLinearInterpolationMaterial::validParams()
       "extrapolation",
       false,
       "Use linear extrapolation to evaluate points that lie outside given data set domain. ");
+  params.declareControllable("scale_factor");
```
Comment (loganharbour) [inline framework/src/materials/PiecewiseLinearInterpolationMaterial.C:40]:
```suggestion
  params.declareControllable("scale_factor");
```

**Review by loganharbour — CHANGES_REQUESTED** [review]

**Review by zachmprince — DISMISSED** [review]

**Review by loganharbour — DISMISSED** [review]

**Comment by snschune** [comment]
How does this fail now? It passed before the rebase.

**Comment by gambka** [comment]
> How does this fail now? It passed before the rebase.

I think the default behavior for the MaterialOutputAux actions which are called when you use output_properties = all has changed to also execute on INITIAL. If I look at the failure, it just looks like the gold file doesn't have values for the material properties on initial and the new one does. I would just say regold your test. See #15433. It looks like there was a legacy flag added so downstream apps have time to update without breaking everything.

**Comment by zachmprince** [comment]
I came to the same conclusion as @gambka (he must of just typed it out faster lol).

**Comment by gambka** [comment]
> I came to the same conclusion as @gambka (he must of just typed it out faster lol).

That is surprising. I'm usually terribly slow at typing responses or describing my own issues and MRs.

**Comment by snschune** [comment]
> Months?! 3.5 weeks :-) 

=> 2 weeks per line of code? 

I think at this point it's a more straight forward way to have my own app with my own piecewise functions. Would certainly be faster and use less of my time.

**Review by zachmprince — APPROVED** [review]

**Review by loganharbour — APPROVED** [review]

**Pull request merged**